File: activity/20200805/junhaolin/897.Increasing_Order_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        if (NULL == root)   
            return root;
        
        if (root->right)
            root->right = increasingBST(root->right);
        
        if (root->left)     {
            TreeNode* newRoot = increasingBST(root->left);
            TreeNode *rightmost = findRightMost(newRoot);
            rightmost->right = root;
            root->left = NULL;
            return newRoot;
        }
        
        return root;
    }
    
    TreeNode* findRightMost(TreeNode* root) {
        if (root->right)
            return findRightMost(root->right);
        else
            return root;
    }
   
};

File: activity/20200805/ninezero/897_increasing_order_search_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func increasingBST(root *TreeNode) *TreeNode {
	var treeValues []int
	bst(root, &treeValues)
	treeNode := &TreeNode{Val: treeValues[0]}
	currentNode := treeNode
	for i := 1; i < len(treeValues); i++ {
		currentNode.Right = &TreeNode{Val: treeValues[i]}
		currentNode = currentNode.Right
	}
	return treeNode
}

func bst(node *TreeNode, treeValues *[]int) {
	if node.Left != nil {
		bst(node.Left, treeValues)
	}
	*treeValues = append(*treeValues, node.Val)
	if node.Right != nil {
		bst(node.Right, treeValues)
	}
}


